Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/overlay/InputOverlayDrawableJoystick.kt

```diff
@@ -0,0 +1,259 @@
+ * @param xControl           The control which the x value of the joystick will be written to.
+ * @param yControl           The control which the y value of the joystick will be written to.
+ */
+class InputOverlayDrawableJoystick(
+    res: Resources,
+    bitmapOuter: Bitmap,
+    bitmapInnerDefault: Bitmap,
+    bitmapInnerPressed: Bitmap,
+    rectOuter: Rect,
+    rectInner: Rect,
+    val legacyId: Int,
+    val xControl: Int,
+    val yControl: Int,
+    private val controllerIndex: Int
+) {
+    var x = 0.0f
+        private set
+    var y = 0.0f
+        private set
+    var trackId = -1
+        private set
+    private var controlPositionX = 0
+    private var controlPositionY = 0
+    private var previousTouchX = 0
+    private var previousTouchY = 0
+    val width: Int
+    val height: Int
+    private var virtBounds: Rect
+    private var origBounds: Rect
+    private var opacity = 0
+    private val outerBitmap: BitmapDrawable
+    private val defaultStateInnerBitmap: BitmapDrawable
+    private val pressedStateInnerBitmap: BitmapDrawable
+    private val boundsBoxBitmap: BitmapDrawable
+    private var pressedState = false
+
+    var bounds: Rect
+        get() = outerBitmap.bounds
+        set(bounds) {
+            outerBitmap.bounds = bounds
+        }
+
+    init {
+        outerBitmap = BitmapDrawable(res, bitmapOuter)
+        defaultStateInnerBitmap = BitmapDrawable(res, bitmapInnerDefault)
+        pressedStateInnerBitmap = BitmapDrawable(res, bitmapInnerPressed)
+        boundsBoxBitmap = BitmapDrawable(res, bitmapOuter)
+        width = bitmapOuter.width
+        height = bitmapOuter.height
+
+        require(!(controllerIndex < 0 || controllerIndex >= 4)) { "controllerIndex must be 0-3" }
+
+        bounds = rectOuter
+        defaultStateInnerBitmap.bounds = rectInner
+        pressedStateInnerBitmap.bounds = rectInner
+        virtBounds = bounds
+        origBounds = outerBitmap.copyBounds()
+        boundsBoxBitmap.alpha = 0
+        boundsBoxBitmap.bounds = virtBounds
+        setInnerBounds()
+    }
+
+    fun draw(canvas: Canvas) {
+        outerBitmap.draw(canvas)
+        currentStateBitmapDrawable.draw(canvas)
+        boundsBoxBitmap.draw(canvas)
+    }
+
+    fun trackEvent(event: MotionEvent): Boolean {
+        val reCenter = BooleanSetting.MAIN_JOYSTICK_REL_CENTER.boolean
+        val action = event.actionMasked
+        val firstPointer = action != MotionEvent.ACTION_POINTER_DOWN &&
+                action != MotionEvent.ACTION_POINTER_UP
+        val pointerIndex = if (firstPointer) 0 else event.actionIndex
+        var pressed = false
+
+        when (action) {
+            MotionEvent.ACTION_DOWN,
+            MotionEvent.ACTION_POINTER_DOWN -> {
+                if (bounds.contains(
+                        event.getX(
+                            pointerIndex
+                        ).toInt(), event.getY(pointerIndex).toInt()
```

Comment:
This kind of stuff kind of hurts to read compared to the Java formatting... But if it has to be formatted that way in the standard way of formatting Kotlin, I guess we'll go with it.

Reply:
That one is pretty bad, but that's only the default formatting. Would this be better?
```suggestion
                if (bounds.contains(
                        event.getX(pointerIndex).toInt(),
                        event.getY(pointerIndex).toInt()
```

Reply:
Yeah, that's a lot better.